src/nvim/option.c

```diff
@@ -1996,6 +1996,414 @@ static void apply_optionset_autocmd(int opt_idx, long opt_flags, long oldval, lo
+  }
+
+  // need to adjust the file name arguments and buffer names.
+  buflist_slash_adjust();
+  alist_slash_adjust();
+  scriptnames_slash_adjust();
+}
+#endif
+
+/// Process the updated 'wrap' option value.
+static void did_set_wrap(void)
+{
+  // If 'wrap' is set, set w_leftcol to zero.
+  if (curwin->w_p_wrap) {
+    curwin->w_leftcol = 0;
+  }
+}
+
+/// Process the updated 'equalalways' option value.
+static void did_set_equalalways(long old_value)
+{
+  if (p_ea && !old_value) {
+    win_equal(curwin, false, 0);
+  }
+}
+
+/// Process the updated 'autochdir' option value.
+static void did_set_autochdir(void)
+{
+  // Change directories when the 'acd' option is set now.
+  do_autochdir();
+}
+
+/// Process the updated 'diff' option value.
+static void did_set_diff(void)
+{
+  // May add or remove the buffer from the list of diff buffers.
+  diff_buf_adjust(curwin);
+  if (foldmethodIsDiff(curwin)) {
+    foldUpdateAll(curwin);
+  }
+}
+
+/// Process the updated 'spell' option value.
+static char *did_set_spell(void)
+{
+  if (curwin->w_p_spell) {
+    return did_set_spelllang(curwin);
+  }
+
+  return NULL;
+}
+
+// Process the updated 'arabic' option value.
+static const char *did_set_arabic(void)
+{
+  const char *errmsg = NULL;
+  if (curwin->w_p_arab) {
+    // 'arabic' is set, handle various sub-settings.
+    if (!p_tbidi) {
+      // set rightleft mode
+      if (!curwin->w_p_rl) {
+        curwin->w_p_rl = true;
+        changed_window_setting();
+      }
+
+      // Enable Arabic shaping (major part of what Arabic requires)
+      if (!p_arshape) {
+        p_arshape = true;
+        redraw_all_later(UPD_NOT_VALID);
+      }
+    }
+
+    // Arabic requires a utf-8 encoding, inform the user if it's not
+    // set.
+    if (strcmp(p_enc, "utf-8") != 0) {
+      static char *w_arabic = N_("W17: Arabic requires UTF-8, do ':set encoding=utf-8'");
+
+      msg_source(HL_ATTR(HLF_W));
+      msg_attr(_(w_arabic), HL_ATTR(HLF_W));
+      set_vim_var_string(VV_WARNINGMSG, _(w_arabic), -1);
+    }
+
+    // set 'delcombine'
+    p_deco = true;
+
+    // Force-set the necessary keymap for arabic.
+    errmsg = set_option_value("keymap", 0L, "arabic", OPT_LOCAL);
+  } else {
+    // 'arabic' is reset, handle various sub-settings.
+    if (!p_tbidi) {
+      // reset rightleft mode
+      if (curwin->w_p_rl) {
+        curwin->w_p_rl = false;
+        changed_window_setting();
+      }
+
+      // 'arabicshape' isn't reset, it is a global option and
+      // another window may still need it "on".
+    }
+
+    // 'delcombine' isn't reset, it is a global option and another
+    // window may still want it "on".
+
+    // Revert to the default keymap
+    curbuf->b_p_iminsert = B_IMODE_NONE;
+    curbuf->b_p_imsearch = B_IMODE_USE_INSERT;
+  }
+
+  return errmsg;
+}
+
+static void did_set_number_relativenumber(void)
+{
+  if (*curwin->w_p_stc != NUL) {
+    // When 'statuscolumn' is set and 'number' or 'relativenumber' is changed:
```

Comment:
If `number` is disabled and then change `relativenumber` then the column width will change. Shouldn't `statuscolumn` be invalidated in these cases?

Reply:
Yes but this is done implicitly because `w_nrwidth_line_count` will always change in `number_width()` when `'relativenumber'` is set/unset.

Resetting them explicitly here is actually only necessary when going from `number` -> `nonumber`

Reply:
ah ok